opensrp-reveal/src/main/java/org/smartregister/reveal/presenter/TaskRegisterFragmentPresenter.java

```diff
@@ -98,17 +100,25 @@ private TaskRegisterFragment getView() {
 
     @Override
     public void onTasksFound(List<TaskDetails> tasks) {
-        this.tasks = tasks;
-        getView().setTaskDetails(tasks);
-        getView().setTotalPatients();
-        getView().hideProgressView();
+        if (recalculateDistance) {//there was a location update when tasks were being retrieved recalculate distance and order
+            interactor.calculateDistanceFromUser(tasks, lastLocation);
+            recalculateDistance = false;
+        } else {
+            this.tasks = tasks;
+            getView().setTaskDetails(tasks);
+            getView().setTotalPatients();
+            getView().hideProgressView();
+        }
 
     }
 
+
     @Override
     public void onLocationChanged(android.location.Location location) {
         if (!location.equals(lastLocation)) {
-            if (lastLocation == null || location.distanceTo(lastLocation) >= Constants.REFRESH_MAP_MINIMUM_DISTANCE) {
+            if (lastLocation == null && tasks == null) {//tasks not yet retrieved from db
+                recalculateDistance = true;
+            } else if (location.distanceTo(lastLocation) >= Constants.REFRESH_MAP_MINIMUM_DISTANCE) {
```

Comment:
Could `location` ever be null?

Reply:
location cannot be null, `onLocationChanged` never returns null locations, `getLastKnownLocation` is the method that can return null location if location services was disabled or before location services is able to retrieve a location

Reply:
ok.